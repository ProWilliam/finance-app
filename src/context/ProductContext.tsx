import React, { createContext, useState, useContext } from 'react';

// Types
import { GlobalProviderProps, ProductContextInterface, Product } from './ProductContext.type';


// Default State Context by products.
const defaultStateProduct = {
  products: [{
    id: '',
    name: ''
  }],
  setProducts: (product: Product[]) => {}

} as ProductContextInterface

const MyContextProducts = createContext<ProductContextInterface>(defaultStateProduct);

export const useMyContextProduct = () => {
  const context = useContext(MyContextProducts);

  if(!context) {
    throw new Error('useMycontext must be used within a MyContextProvider');
  }

  return context;
}

export const ProductProvider: React.FC<GlobalProviderProps> = ({ children }) => {
  const [ products, setProducts ] = useState<Product[]>([]);

  return (
    <MyContextProducts.Provider value={{ products, setProducts }}>
      {children}
    </MyContextProducts.Provider>
  );
};