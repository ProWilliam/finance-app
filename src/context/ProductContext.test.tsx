import React from 'react';
import renderer, { act, ReactTestRenderer } from 'react-test-renderer';
import { Text, View, TouchableOpacity } from 'react-native';

import { ProductProvider, useMyContextProduct } from './ProductContext';

// Helper component to access the context values in tests
const TestComponent: React.FC = () => {
  const { products, setProducts } = useMyContextProduct();
  const listProduct = [
    { id: '2', name: 'Product 2' },
    { id: '3', name: 'Product 3' }
  ]
  return (
    <View>
      {products.map((product) => (
        <Text key={product.id} testID="product-item">{product.name}</Text>
      ))}
      <TouchableOpacity onPress={() => setProducts(listProduct)} testID="update-button">
        Update Products
      </TouchableOpacity>
    </View>
  );
};

describe('<ProductContext />', () => {

  let component: ReactTestRenderer;  

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const tree = renderer.create(
      <ProductProvider>
        <TestComponent />
      </ProductProvider>).toJSON();

    expect(tree).toMatchSnapshot();
  });
  
  it('provides default empty products array', () => {
    
    act(() => {
      component = renderer.create(
        <ProductProvider>
          <TestComponent />
        </ProductProvider>
      );
    });

    
    const productItems = component.root.findAllByProps({ testID: 'product-item' });
    expect(productItems.length).toBe(0); // Expect the default state to have an empty products array
    
  });

  it('updates products in context when setProducts is called', () => {
    
    act(() => {
      component = renderer.create(
        <ProductProvider>
          <TestComponent />
        </ProductProvider>
      );
    });

    
    // Simulate button click to update products
    act(() => {
      component.root.findByProps({ testID: 'update-button' }).props.onPress();
    });

    const productItems = component.root.findAllByType(Text);
    
    expect(productItems.length).toBe(2);
    expect(productItems[0].props.children).toContain('Product 2');
    expect(productItems[1].props.children).toContain('Product 3');
    
  });

  it('initializes with the default state', () => {
    
    act(() => {
      component = renderer.create(
        <ProductProvider>
          <TestComponent />
        </ProductProvider>
      );
    });

    
    const productItems = component.root.findAllByProps(Text);
    expect(productItems.length).toBe(0);
    
  });

  it('does not affect context state when ProductProvider is re-rendered', () => {
    
    act(() => {
      component = renderer.create(
        <ProductProvider>
          <TestComponent />
        </ProductProvider>
      );
    });

    
    act(() => {
      component.update(
        <ProductProvider>
          <TestComponent />
        </ProductProvider>
      );
    });

    const productItems = component.root.findAllByType(Text);
    expect(productItems.length).toBe(0); // Should still be empty after re-render
    
  });
  
});
