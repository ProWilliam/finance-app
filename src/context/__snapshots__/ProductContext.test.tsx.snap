// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductContext /> renders correctly 1`] = `
<View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="update-button"
  >
    Update Products
  </View>
</View>
`;
